Add how to verify l2 state on l1 page (#1067)
OffchainLabs/arbitrum-docs: docs/build-decentralized-apps/how-to-get-l2block-on-l1.mdx

---
title: 'How to verify child chain state on parent chain'
description: Learn how to verify child chain state on its parent chain
user_story: As a developer, I want to understand how to verify child chain state on parent chain.
content_type: how-to
---

Arbitrum implements a **fraud proof** system that ensures that the state of any given child chain is safely maintained by its parent chain. In this system, a validator is responsible for periodically transmitting information about the child chain's state to its parent chain.
This information is included in "rollup blocks", which we refer to as **RBlocks** throughout our docs. See [Inside Arbitrum Nitro](../inside-arbitrum-nitro/inside-arbitrum-nitro.mdx#arbitrum-rollup-protocol) to learn more about RBlocks.

These RBlocks are effectively _assertions about the impact that a series of child chain blocks (containing transactions) ought to have on its parent chain's state_. When received by the parent chain, these RBlocks are decoded by a contract hosted by the parent chain (an "on-chain contract"); this information can then be optionally relayed to off-chain tools for arbitrary purposes (for example, off-chain validation).

> Before we begin, we will introduce the key component: rblock, assertion and send roots.

Reviewer: ```suggestion
Before we begin, we will introduce the key components: RBlock, assertion, and send roots.
```

> # Rblock and Assertion

Reviewer: ```suggestion
# RBlock and assertion
```

> The rollup contract contains a series of components used to maintain the operation of the layer2 network, including rblocks.

Reviewer: ```suggestion
The rollup contract contains a series of components used to maintain the operation of the child chain, including RBlocks.
```

    bytes32 stateHash;
    // Hash of the data that can be challenged
    bytes32 challengeHash;
    // Hash of the data that will be committed if this node is confirmed
    bytes32 confirmData;
    // Index of the node previous to this one
    uint64 prevNum;
    // Deadline at which this node can be confirmed
    uint64 deadlineBlock;
    // Deadline at which a child of this node can be confirmed
    uint64 noChildConfirmedBeforeBlock;
    // Number of stakers staked on this node. This includes real stakers and zombies
    uint64 stakerCount;
    // Number of stakers staked on a child node. This includes real stakers and zombies
    uint64 childStakerCount;
    // This value starts at zero and is set to a value when the first child is created. After that it is constant until the node is destroyed or the owner destroys pending nodes
    uint64 firstChildBlock;
    // The number of the latest child of this node to be created
    uint64 latestChildNumber;
    // The block number when this node was created
    uint64 createdAtBlock;
    // A hash of all the data needed to determine this node's validity, to protect against reorgs
    bytes32 nodeHash;
}
```

> When creating a new rblock, a new assertion will be made too:

Reviewer: ```suggestion
When creating a new RBlock, a new assertion will be made too:
```

```
struct Assertion {
    ExecutionState beforeState;
    ExecutionState afterState;
    uint64 numBlocks;
}
```

> As we can see above, an rblock has a series of field, they are useful when validators try to challenge or confirm this rblock.

Reviewer: ```suggestion
As we can see above, an RBlock has a series of fields. They are useful when validators try to challenge or confirm this RBlock.
```

> What we can use here is the `confirmData`, the `confirmData` is the keccak256 hash of child chain block Hash and sendRoot.

Reviewer: ```suggestion
What we can use here is the `confirmData`, the `confirmData` is the keccak256 hash of child chain block hash and sendRoot.
```

As for Assertion, it has 2 `ExecutionState` which is the start state and the end state of this assertion, and `ExecutionState` contains the information about child chain blockhash and related sendroot, so we can extract `blockhash` from there.

# Send roots

> The send root mapping is stored in the outbox contract. This mapping is used to store the Merkle root of each batch of child chain -> parent chain transactions called send root and its corresponding child chain block hash.

Reviewer: ```suggestion
The send root mapping is stored in the outbox contract. This mapping is used to store the Merkle root of each batch of child chain -> parent chain transactions, called a send root, and its corresponding child chain block hash.
```

> When an rblock is confirmed, the corresponding send root will be recorded to outbox contract from rollup contract so when an user wants to triger the child chain -> parent chain transaction on parent chain the transaction requests can be verified.

Reviewer: ```suggestion
When an RBlock is confirmed, the corresponding send root will be recorded to the outbox contract from the rollup contract, so when a user wants to trigger the child chain -> parent chain transaction on parent chain, the transaction requests can be verified.
```

```
mapping(bytes32 => bytes32) public roots; // maps root hashes => child chain block hash
```

> This mapping will save the `blockhash`, so we can get the child chain blockhash from the outbox contract too.

Reviewer: ```suggestion
This mapping will save the `blockhash`, so we can get the child chain block hash from the outbox contract too.
```

# Verify child chain state on parent chain

> Assume that there is a contract called `foo` on child chain, and its contract address is `fooAddress`, now we want to prove its state value at storage `slot`.

Reviewer: ```suggestion
Assume that there is a contract called `foo` on child chain, and its contract address is `fooAddress`. Now we want to prove its state value at storage `slot`.
```

To verify the state, we need a Merkle Trie Verifier contract, one example is [Lib_MerkleTrie.sol](https://github.com/ethereum-optimism/optimism-legacy/blob/8205f678b7b4ac4625c2afe351b9c82ffaa2e795/packages/contracts/contracts/libraries/trie/Lib_MerkleTrie.sol).

## 1. How to verify a confirmed child chain block hash

> For the security of verification, we will use the latest confirmation instead of the latest proposed rblock for verification:

Reviewer: ```suggestion
For the security of verification, we will use the latest confirmation instead of the latest proposed RBlock for verification:
```

> - Obtain the latest confirmed rblock from rollup contract: `nodeIndex = rollup.latestConfirmed()`, this step will return the corresponding rblock number: `nodeIndex`

Reviewer: ```suggestion
- Obtain the latest confirmed RBlock from rollup contract: `nodeIndex = rollup.latestConfirmed()`, this step will return the corresponding RBlock number: `nodeIndex`
```

> - Filter the event with the obtained rblock number: `nodeEvent = NodeCreated(nodeIndex)`, and get the corresponding assertion information: `assertion = nodeEvents[0].args.assertion`

Reviewer: ```suggestion
- Filter the event with the obtained RBlock number: `nodeEvent = NodeCreated(nodeIndex)`, and get the corresponding assertion information: `assertion = nodeEvents[0].args.assertion`
```

- Fetch blockhash via `blockhash = GlobalStateLib.getBlockHash(assertion.afterState.globalState)` (As mentioned above, you can also get the block hash from the outbox contract)
- Fetch sendRoot via `sendRoot = GlobalStateLib.getSendRoot(assertion.afterState.globalState)`

> - After getting the blockhash, we need to compare it with the confirmdata in rblock, to get the confirm data: `confirmdata = keccak256(solidityPack(['bytes32','bytes32'], [blockHash, sendRoot]))`

Reviewer: ```suggestion
- After getting the blockhash, we need to compare it with the confirmData in the RBlock. To get the confirm data: `confirmdata = keccak256(solidityPack(['bytes32','bytes32'], [blockHash, sendRoot]))`
```

> - Get the corresponding rblock: `rblock = rollup.getNode(nodeIndex)`

Reviewer: ```suggestion
- Get the corresponding RBlock: `rblock = rollup.getNode(nodeIndex)`
```

- Compare if they have the same value: `rblock.confirmData == confirmdata`

> ## 2. Proof the state root belong to the child chain block hash by supplying the blockheader

Reviewer: ```suggestion
## 2. Prove the state root belongs to the child chain block hash by supplying the block header
```

After we obtain the block hash, we can obtain the corresponding block information from child chain provider: `l2blockRaw = eth_getBlockByHash (blockhash)`

Next, we need to manually derive blockhash by hashing block header fields.

```
blockarray = [
    l2blockRaw.parentHash,
    l2blockRaw.sha3Uncles,
    l2blockRaw.miner,
    l2blockRaw.stateroot,
    l2blockRaw.transactionsRoot,
    l2blockRaw.receiptsRoot,
    l2blockRaw.logsBloom,
    BigNumber.from(l2blockRaw.difficulty).toHexString(),
    BigNumber.from(l2blockRaw.number).toHexString(),
    BigNumber.from(l2blockRaw.gasLimit).toHexString(),
    BigNumber.from(l2blockRaw.gasUsed).toHexString(),
    BigNumber.from(l2blockRaw.timestamp).toHexString(),
    l2blockRaw.extraData,
    l2blockRaw.mixHash,
    l2blockRaw.nonce,
    BigNumber.from(l2blockRaw.baseFeePerGas).toHexString(),
    ]
```

- Calculate the block hash to verify whether the information in the obtained block is correct: `calculated_blockhash = keccak256(RLP.encode(blockarray))`

> - Verify whether the block hash is same with what we got from assertion or outbox contract: `calculated_blockhash === blockHash`

Reviewer: ```suggestion
- Verify whether the block hash is the same as what we got from assertion or outbox contract: `calculated_blockhash === blockHash`
```

> If it is same, it can be used to prove that the information in the block header, especially the stateroot, is correct.

Reviewer: ```suggestion
If it is the same, it can be used to prove that the information in the block header, especially the state root, is correct.
```

> ## 3. Proof the account storage inside the state root

Reviewer: ```suggestion
## 3. Prove the account storage inside the state root
```

After we obtain the correct state root, we can continue to verify the storage slot.

- First, we need to obtain the proof of the corresponding state root from child chain:

```
proof = l2provider.send('eth_getProof', [
    fooAddress,
    [slot],
    {blockHash}
  ]);
```

- Get account proof: `accountProof = RLP.encode(proof.accountProof)`
- Get proofKey: `proofKey = ethers.utils.keccak256(fooAddress)`
- Call the verifier contract to verify:

```
[acctExists, acctEncoded] = verifier.get(
    proofKey, accountProof, stateroot
  )
```

- Check for equality: `acctExists == true`

> ## 4. Proof the storage slot is in the account root

Reviewer: ```suggestion
## 4. Prove the storage slot is in the account root
```

- Get storage root: `storageRoot = RLP.decode(acctEncoded)[2]`
- Get storage slot key: `slotKey = ethers.utils.keccak256(slot)`
- Get storageProof: `storageProof = ethers.utils.RLP.encode((proof.storageProof as any[]).filter((x)=>x.key===slot)[0].proof)`

> - Call the merkle verifier contract to verify:

Reviewer: ```suggestion
- Call the Merkle verifier contract to verify:
```

```
const [storageExists, storageEncoded] = await verifier.get(
    slotKey, storageProof, storageRoot
  )
```

- Check for equality: `storageExists == true`
- Obtain the value of the storage as `slot`: `storageValue = ethers.utils.RLP.decode(storageEncoded)`

> Then we can successfuly prove and get a certain state value at a specific block height on child chain through parent chain.

Reviewer: ```suggestion
Then we can successfully prove and get a certain state value at a specific block height on child chain through parent chain.
```

### Let's check this value on child chain directly

> - Call child chain rpc provider to get the value of the corresponding block number: `actualValue = l2provider.getStorageAt(fooAddress, slot, l2blockRaw.number)`

Reviewer: ```suggestion
- Call child chain RPC provider to get the value of the corresponding block number: `actualValue = l2provider.getStorageAt(fooAddress, slot, l2blockRaw.number)`
```

---

> title: 'How to verify child chain state on parent chain'

Reviewer: ```suggestion
title: 'How to verify child chain state on the parent chain'
```

description: Learn how to verify child chain state on its parent chain

> user_story: As a developer, I want to understand how to verify child chain state on parent chain.

Reviewer: ```suggestion
user_story: As a developer, I want to understand how to verify child chain state on the parent chain.
```

```solidity
struct AssertionNode {
    // Block when the first child of this assertion was created
    uint64 firstChildBlock;
    // Block when the second child was created (non-zero => assertion was challenged)
    uint64 secondChildBlock;
    // The block number when this assertion was created
    uint64 createdAtBlock;
    // True if this assertion is the first child of its prev
    bool isFirstChild;
    // Status of the assertion: NoAssertion / Pending / Confirmed
    AssertionStatus status;
    // Hash of the environment config at creation time (e.g. wasmModuleRoot)
    bytes32 configHash;
}
```

An assertion's validity is bound to its **assertion hash**; the struct above does not store the claimed state directly.

An assertion is created together with its **before** and **after** execution states, packaged as `AssertionInputs`:

```solidity
struct AssertionInputs {
    BeforeStateData beforeStateData;
    AssertionState beforeState;
    AssertionState afterState;
}

struct AssertionState {
    GlobalState globalState;
    MachineStatus machineStatus;
    bytes32 endHistoryRoot;
}
```

> The piece we care about is `afterState.globalState`: it contains the child chain **block hash** and **send root** that this assertion claims. We can extract them with the `GlobalStateLib` helpers `getBlockHash()` and `getSendRoot()`.

Reviewer: ```suggestion
The key field is `afterState.globalState`: it contains the child chain **block hash** and **send root** that this assertion claims. We can extract them with the `GlobalStateLib` helpers `getBlockHash()` and `getSendRoot()`.
```

The assertion hash itself authenticates these values — it is computed as:

```solidity
assertionHash = keccak256(abi.encodePacked(
    parentAssertionHash,
    afterState.hash(),   // keccak256(abi.encode(afterState))
    inboxAcc
))
```

## Send roots

The send root mapping is stored in the `Outbox` contract. It maps the Merkle root of each batch of child chain -> parent chain messages (the _send root_) to its corresponding child chain block hash.

When an assertion is confirmed, the rollup contract records the send root to the outbox so that, when a user later triggers a child chain -> parent chain message on the parent chain, the request can be verified.

```solidity
mapping(bytes32 => bytes32) public roots; // maps root hashes => child chain block hash
```

Because this mapping stores the block hash, you can also recover the child chain block hash directly from the outbox.

## Verify child chain state on parent chain

Assume there is a contract called `foo` on the child chain at address `fooAddress`, and we want to prove its storage value at `slot`.

> To verify the state, we need a Merkle Trie verifier contract — for example a `Lib_MerkleTrie`-style library that exposes a `get(key, proof, root)` function.

Reviewer: ```suggestion
To verify the state, we need a Merkle Trie verifier contract — for example, a `Lib_MerkleTrie`-style library that exposes a `get(key, proof, root)` function.
```

:::note Why we can trust an untrusted child chain RPC

The steps below query a child chain RPC several times (`eth_getBlockByHash`, `eth_getProof`). You do _not_ have to trust that RPC. Every value it returns is checked against the trust anchor established in step 1 — the block hash and send root taken from the assertion **confirmed on the parent chain**:

- The block header is only accepted if its RLP hash reproduces the confirmed block hash.
- The account and storage values are only accepted if their Merkle proofs verify against the state root inside that header.

> Because each link is bound to the confirmed parent chain commitment by a cryptographic hash, a malicious or buggy RPC **cannot forge data that passes verification**. The worst it can do is return wrong or missing data, which makes verification _fail_ (a denial of service) — it can never produce a false positive. This one-directional guarantee is what makes the whole procedure sound.

Reviewer: ```suggestion
Because each link is bound to the confirmed parent chain commitment by a cryptographic hash, a malicious or buggy RPC **cannot forge data that passes verification**. The worst it can do is return wrong or missing data, which makes verification _fail_ (a denial of service) — it can never produce a false positive. This one-way guarantee is what makes the whole procedure sound.
```

:::

### 1. Get a confirmed child chain block hash

For security, we use the latest _confirmed_ assertion rather than the latest _proposed_ one. The `AssertionConfirmed` event emits the block hash and send root directly:

- Get the latest confirmed assertion hash: `assertionHash = rollup.latestConfirmed()`, which returns a `bytes32` assertion hash.
- Query the confirmation event for that hash: `AssertionConfirmed(bytes32 indexed assertionHash, bytes32 blockHash, bytes32 sendRoot)`. The event gives you `blockHash` and `sendRoot` directly.
- (Optional) You can cross-check by reading the global state from the `AssertionCreated` event for the same hash and extracting `blockHash = GlobalStateLib.getBlockHash(assertion.afterState.globalState)` and `sendRoot = GlobalStateLib.getSendRoot(assertion.afterState.globalState)`.
- (Optional) You can also look the block hash up in the outbox: `roots[sendRoot]`.

### 2. Prove the state root belongs to the block hash, using the block header

> Having the block hash, fetch the corresponding block from the child chain provider: `l2blockRaw = eth_getBlockByHash(blockHash)`.

Reviewer: ```suggestion
With the block hash, fetch the corresponding block from the child chain provider: `l2blockRaw = eth_getBlockByHash(blockHash)`.
```

Then re-derive the block hash by RLP-encoding and hashing the header fields:

> ```

Reviewer: ```suggestion
```shell
```

blockarray = [
  l2blockRaw.parentHash,
  l2blockRaw.sha3Uncles,
  l2blockRaw.miner,
  l2blockRaw.stateRoot,
  l2blockRaw.transactionsRoot,
  l2blockRaw.receiptsRoot,
  l2blockRaw.logsBloom,
  BigNumber.from(l2blockRaw.difficulty).toHexString(),
  BigNumber.from(l2blockRaw.number).toHexString(),
  BigNumber.from(l2blockRaw.gasLimit).toHexString(),
  BigNumber.from(l2blockRaw.gasUsed).toHexString(),
  BigNumber.from(l2blockRaw.timestamp).toHexString(),
  l2blockRaw.extraData,
  l2blockRaw.mixHash,
  l2blockRaw.nonce,
  BigNumber.from(l2blockRaw.baseFeePerGas).toHexString(),
];
```

- Compute `calculated_blockhash = keccak256(RLP.encode(blockarray))`.

> - Check it matches the value from step 1: `calculated_blockhash === blockHash`.

Reviewer: ```suggestion
- Check that it matches the value from step 1: `calculated_blockhash === blockHash`.
```
